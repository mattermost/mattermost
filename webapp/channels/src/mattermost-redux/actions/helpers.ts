// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ActionTypes} from 'mattermost-redux/action_types';
import {GenericAction} from 'mattermost-redux/types/actions';

export function getSharedChannels(): GenericAction {
    return {
        type: ActionTypes.GET_SHARED_CHANNELS,
    };
}

export function receivedSharedChannelsWithRemotes(sharedChannelsWithRemotes: any[]): GenericAction {
    return {
        type: ActionTypes.RECEIVED_SHARED_CHANNELS_WITH_REMOTES,
        data: sharedChannelsWithRemotes,
    };
}