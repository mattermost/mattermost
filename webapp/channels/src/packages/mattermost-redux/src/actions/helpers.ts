// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AnyAction} from 'redux';

import type {ServerError} from '@mattermost/types/errors';
import type {GlobalState} from '@mattermost/types/store';

import {UserTypes} from 'mattermost-redux/action_types';
import {Client4} from 'mattermost-redux/client';
import type {DispatchFunc, GetStateFunc, ActionFuncAsync} from 'mattermost-redux/types/actions';

import {logError} from './errors';

type ActionType = string;
const HTTP_UNAUTHORIZED = 401;
export function forceLogoutIfNecessary(err: ServerError, dispatch: DispatchFunc<AnyAction>, getState: GetStateFunc) {
    const {currentUserId} = getState().entities.users;

    if ('status_code' in err && err.status_code === HTTP_UNAUTHORIZED && err.url && err.url.indexOf('/login') === -1 && currentUserId) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

function dispatcher(type: ActionType, data: any, dispatch: DispatchFunc<AnyAction>) {
    if (type.indexOf('SUCCESS') === -1) { // we don't want to pass the data for the request types
        dispatch(requestSuccess(type, data));
    } else {
        dispatch(requestData(type));
    }
}

export function requestData(type: ActionType) {
    return {
        type,
        data: null,
    };
}

export function requestSuccess(type: ActionType, data: any) {
    return {
        type,
        data,
    };
}

export function requestFailure(type: ActionType, error: ServerError): any {
    return {
        type,
        error,
    };
}

/**
 * Returns an ActionFunc which calls a specfied (client) function and
 * dispatches the specifed actions on request, success or failure.
 *
 * @export
 * @param {Object} obj                                       an object for destructirung required properties
 * @param {() => Promise<mixed>} obj.clientFunc              clientFunc to execute
 * @param {ActionType} obj.onRequest                         ActionType to dispatch on request
 * @param {(ActionType | Array<ActionType>)} obj.onSuccess   ActionType to dispatch on success
 * @param {ActionType} obj.onFailure                         ActionType to dispatch on failure
 * @param {...Array<any>} obj.params
 * @returns {ActionFunc} ActionFunc
 */

export function bindClientFunc<Func extends(...args: any[]) => Promise<any>>({
    clientFunc,
    onRequest,
    onSuccess,
    onFailure,
    params,
}: {
    clientFunc: Func;
    onRequest?: ActionType;
    onSuccess?: ActionType | ActionType[];
    onFailure?: ActionType;
    params?: Parameters<Func>;
}): ActionFuncAsync<Awaited<ReturnType<Func>>, GlobalState, AnyAction> {
    return async (dispatch, getState) => {
        if (onRequest) {
            dispatch(requestData(onRequest));
        }

        let data: Awaited<ReturnType<Func>>;
        try {
            if (params) {
                data = await clientFunc(...params);
            } else {
                data = await clientFunc();
            }
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            if (onFailure) {
                dispatch(requestFailure(onFailure, error));
            }
            dispatch(logError(error));
            return {error};
        }

        if (Array.isArray(onSuccess)) {
            onSuccess.forEach((s) => {
                dispatcher(s, data, dispatch);
            });
        } else if (onSuccess) {
            dispatcher(onSuccess, data, dispatch);
        }

        return {data};
    };
}

// Debounce function based on underscores modified to use es6 and a cb

export function debounce(func: (...args: any) => unknown, wait: number, immediate?: boolean, cb?: () => unknown) {
    let timeout: NodeJS.Timeout|null;
    return function fx(...args: any[]) {
        const runLater = () => {
            timeout = null;
            if (!immediate) {
                Reflect.apply(func, null, args);
                if (cb) {
                    cb();
                }
            }
        };
        const callNow = immediate && !timeout;
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(runLater, wait);
        if (callNow) {
            Reflect.apply(func, null, args);
            if (cb) {
                cb();
            }
        }
    };
}

export class FormattedError extends Error {
    intl: {
        id: string;
        defaultMessage: string;
        values: any;
    };

    constructor(id: string, defaultMessage: string, values: any = {}) {
        super(defaultMessage);
        this.intl = {
            id,
            defaultMessage,
            values,
        };
    }
}

