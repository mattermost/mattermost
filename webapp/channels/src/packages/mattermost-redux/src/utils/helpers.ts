// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import shallowEqual from 'shallow-equals';

import {createSelectorCreator, defaultMemoize} from 'mattermost-redux/selectors/create_selector';

// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
export function memoizeResult<F extends Function>(func: F, measure: Function | undefined = undefined): F {
    let lastArgs: IArguments|null = null;
    let lastResult: any = null;

    // we reference arguments instead of spreading them for performance reasons
    return function memoizedFunc() {
        if (!shallowEqual(lastArgs, arguments)) { //eslint-disable-line prefer-rest-params
            //eslint-disable-line prefer-rest-params
            // apply arguments instead of spreading for performance.
            const result = Reflect.apply(func, null, arguments); //eslint-disable-line prefer-rest-params
            if (!shallowEqual(lastResult, result)) {
                lastResult = result;
            }
        }

        if (measure) {
            measure();
        }

        lastArgs = arguments; //eslint-disable-line prefer-rest-params
        return lastResult;
    } as unknown as F;
}

// Use this selector when you want a shallow comparison of the arguments and you want to memoize the result
// try and use this only when your selector returns an array of ids
export const createIdsSelector = createSelectorCreator(memoizeResult);

// Use this selector when you want a shallow comparison of the arguments and you don't need to memoize the result
export const createShallowSelector = createSelectorCreator(defaultMemoize, shallowEqual as any);

// isMinimumServerVersion will return true if currentVersion is equal to higher or than the
// the provided minimum version. A non-equal major version will ignore minor and dot
// versions, and a non-equal minor version will ignore dot version.
// currentVersion is a string, e.g '4.6.0'
// minMajorVersion, minMinorVersion, minDotVersion are integers
export const isMinimumServerVersion = (currentVersion: string, minMajorVersion = 0, minMinorVersion = 0, minDotVersion = 0): boolean => {
    if (!currentVersion || typeof currentVersion !== 'string') {
        return false;
    }

    const split = currentVersion.split('.');

    const major = parseInt(split[0], 10);
    const minor = parseInt(split[1] || '0', 10);
    const dot = parseInt(split[2] || '0', 10);

    if (major > minMajorVersion) {
        return true;
    }
    if (major < minMajorVersion) {
        return false;
    }

    // Major version is equal, check minor
    if (minor > minMinorVersion) {
        return true;
    }
    if (minor < minMinorVersion) {
        return false;
    }

    // Minor version is equal, check dot
    if (dot > minDotVersion) {
        return true;
    }
    if (dot < minDotVersion) {
        return false;
    }

    // Dot version is equal
    return true;
};

// Generates a RFC-4122 version 4 compliant globally unique identifier.
export function generateId(): string {
    // implementation taken from http://stackoverflow.com/a/2117523
    let id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
    id = id.replace(/[xy]/g, (c) => {
        const r = Math.floor(Math.random() * 16);
        let v;

        if (c === 'x') {
            v = r;
        } else {
            // eslint-disable-next-line no-mixed-operators
            v = r & 0x3 | 0x8;
        }

        return v.toString(16);
    });
    return id;
}

export function isEmail(email: string): boolean {
    // writing a regex to match all valid email addresses is really, really hard. (see http://stackoverflow.com/a/201378)
    // this regex ensures:
    // - at least one character that is not a space, comma, or @ symbol
    // - followed by a single @ symbol
    // - followed by at least one character that is not a space, comma, or @ symbol
    // this prevents <Outlook Style> outlook.style@domain.com addresses and multiple comma-separated addresses from being accepted
    return (/^[^ ,@]+@[^ ,@]+$/).test(email);
}
