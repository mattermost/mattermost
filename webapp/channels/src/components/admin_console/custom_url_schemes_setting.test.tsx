// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';

import CustomURLSchemesSetting from './custom_url_schemes_setting';

describe('components/AdminConsole/CustomUrlSchemeSetting', () => {
    const baseProps = {
        id: 'MySetting',
        value: ['git', 'smtp'],
        onChange: jest.fn(),
        disabled: false,
        setByEnv: false,
    };

    describe('initial state', () => {
        test('with no items', () => {
            const props = {
                ...baseProps,
                value: [],
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );
            expect(wrapper).toMatchSnapshot();

            expect(wrapper.state('value')).toEqual('');
        });

        test('with one item', () => {
            const props = {
                ...baseProps,
                value: ['git'],
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );
            expect(wrapper).toMatchSnapshot();

            expect(wrapper.state('value')).toEqual('git');
        });

        test('with multiple items', () => {
            const props = {
                ...baseProps,
                value: ['git', 'smtp', 'steam'],
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );
            expect(wrapper).toMatchSnapshot();

            expect(wrapper.state('value')).toEqual('git,smtp,steam');
        });
    });

    describe('onChange', () => {
        test('called on change to empty', () => {
            const props = {
                ...baseProps,
                onChange: jest.fn(),
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );

            wrapper.find('input').simulate('change', {target: {value: ''}});

            expect(props.onChange).toBeCalledWith(baseProps.id, []);
        });

        test('called on change to one item', () => {
            const props = {
                ...baseProps,
                onChange: jest.fn(),
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );

            wrapper.find('input').simulate('change', {target: {value: '  steam  '}});

            expect(props.onChange).toBeCalledWith(baseProps.id, ['steam']);
        });

        test('called on change to two items', () => {
            const props = {
                ...baseProps,
                onChange: jest.fn(),
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );

            wrapper.find('input').simulate('change', {target: {value: 'steam, git'}});

            expect(props.onChange).toBeCalledWith(baseProps.id, ['steam', 'git']);
        });

        test('called on change to more items', () => {
            const props = {
                ...baseProps,
                onChange: jest.fn(),
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );

            wrapper.find('input').simulate('change', {target: {value: 'ts3server, smtp, ms-excel'}});

            expect(props.onChange).toBeCalledWith(baseProps.id, ['ts3server', 'smtp', 'ms-excel']);
        });

        test('called on change with extra commas', () => {
            const props = {
                ...baseProps,
                onChange: jest.fn(),
            };

            const wrapper = mountWithIntl(
                <CustomURLSchemesSetting {...props}/>,
            );

            wrapper.find('input').simulate('change', {target: {value: ',,,,,chrome,,,,ms-excel,,'}});

            expect(props.onChange).toBeCalledWith(baseProps.id, ['chrome', 'ms-excel']);
        });
    });

    test('renders properly when disabled', () => {
        const props = {
            ...baseProps,
            disabled: true,
        };

        const wrapper = mountWithIntl(
            <CustomURLSchemesSetting {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('renders properly when set by environment variable', () => {
        const props = {
            ...baseProps,
            setByEnv: true,
        };

        const wrapper = mountWithIntl(
            <CustomURLSchemesSetting {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });
});
