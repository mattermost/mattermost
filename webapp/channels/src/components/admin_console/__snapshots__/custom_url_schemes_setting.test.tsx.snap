// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AdminConsole/CustomUrlSchemeSetting initial state with multiple items 1`] = `
<CustomURLSchemesSetting
  disabled={false}
  id="MySetting"
  onChange={[MockFunction]}
  setByEnv={false}
  value={
    Array [
      "git",
      "smtp",
      "steam",
    ]
  }
>
  <Memo(Settings)
    helpText={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
        id="admin.customization.customUrlSchemesDesc"
      />
    }
    inputId="MySetting"
    label={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Custom URL Schemes:"
        id="admin.customization.customUrlSchemes"
      />
    }
    setByEnv={false}
  >
    <div
      className="form-group"
      data-testid="MySetting"
    >
      <label
        className="control-label col-sm-4"
        htmlFor="MySetting"
      >
        <FormattedMessage
          defaultMessage="Custom URL Schemes:"
          id="admin.customization.customUrlSchemes"
        >
          <span>
            Custom URL Schemes:
          </span>
        </FormattedMessage>
      </label>
      <div
        className="col-sm-8"
      >
        <LocalizedPlaceholderInput
          className="form-control"
          disabled={false}
          id="MySetting"
          onChange={[Function]}
          placeholder={
            Object {
              "defaultMessage": "E.g.: \\"git,smtp\\"",
              "id": "admin.customization.customUrlSchemesPlaceholder",
            }
          }
          type="text"
          value="git,smtp,steam"
        >
          <input
            className="form-control"
            disabled={false}
            id="MySetting"
            onChange={[Function]}
            placeholder="E.g.: \\"git,smtp\\""
            type="text"
            value="git,smtp,steam"
          />
        </LocalizedPlaceholderInput>
        <div
          className="help-text"
          data-testid="MySettinghelp-text"
        >
          <FormattedMessage
            defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
            id="admin.customization.customUrlSchemesDesc"
          >
            <span>
              Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: "http", "https", "ftp", "tel", and "mailto".
            </span>
          </FormattedMessage>
        </div>
      </div>
    </div>
  </Memo(Settings)>
</CustomURLSchemesSetting>
`;

exports[`components/AdminConsole/CustomUrlSchemeSetting initial state with no items 1`] = `
<CustomURLSchemesSetting
  disabled={false}
  id="MySetting"
  onChange={[MockFunction]}
  setByEnv={false}
  value={Array []}
>
  <Memo(Settings)
    helpText={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
        id="admin.customization.customUrlSchemesDesc"
      />
    }
    inputId="MySetting"
    label={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Custom URL Schemes:"
        id="admin.customization.customUrlSchemes"
      />
    }
    setByEnv={false}
  >
    <div
      className="form-group"
      data-testid="MySetting"
    >
      <label
        className="control-label col-sm-4"
        htmlFor="MySetting"
      >
        <FormattedMessage
          defaultMessage="Custom URL Schemes:"
          id="admin.customization.customUrlSchemes"
        >
          <span>
            Custom URL Schemes:
          </span>
        </FormattedMessage>
      </label>
      <div
        className="col-sm-8"
      >
        <LocalizedPlaceholderInput
          className="form-control"
          disabled={false}
          id="MySetting"
          onChange={[Function]}
          placeholder={
            Object {
              "defaultMessage": "E.g.: \\"git,smtp\\"",
              "id": "admin.customization.customUrlSchemesPlaceholder",
            }
          }
          type="text"
          value=""
        >
          <input
            className="form-control"
            disabled={false}
            id="MySetting"
            onChange={[Function]}
            placeholder="E.g.: \\"git,smtp\\""
            type="text"
            value=""
          />
        </LocalizedPlaceholderInput>
        <div
          className="help-text"
          data-testid="MySettinghelp-text"
        >
          <FormattedMessage
            defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
            id="admin.customization.customUrlSchemesDesc"
          >
            <span>
              Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: "http", "https", "ftp", "tel", and "mailto".
            </span>
          </FormattedMessage>
        </div>
      </div>
    </div>
  </Memo(Settings)>
</CustomURLSchemesSetting>
`;

exports[`components/AdminConsole/CustomUrlSchemeSetting initial state with one item 1`] = `
<CustomURLSchemesSetting
  disabled={false}
  id="MySetting"
  onChange={[MockFunction]}
  setByEnv={false}
  value={
    Array [
      "git",
    ]
  }
>
  <Memo(Settings)
    helpText={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
        id="admin.customization.customUrlSchemesDesc"
      />
    }
    inputId="MySetting"
    label={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Custom URL Schemes:"
        id="admin.customization.customUrlSchemes"
      />
    }
    setByEnv={false}
  >
    <div
      className="form-group"
      data-testid="MySetting"
    >
      <label
        className="control-label col-sm-4"
        htmlFor="MySetting"
      >
        <FormattedMessage
          defaultMessage="Custom URL Schemes:"
          id="admin.customization.customUrlSchemes"
        >
          <span>
            Custom URL Schemes:
          </span>
        </FormattedMessage>
      </label>
      <div
        className="col-sm-8"
      >
        <LocalizedPlaceholderInput
          className="form-control"
          disabled={false}
          id="MySetting"
          onChange={[Function]}
          placeholder={
            Object {
              "defaultMessage": "E.g.: \\"git,smtp\\"",
              "id": "admin.customization.customUrlSchemesPlaceholder",
            }
          }
          type="text"
          value="git"
        >
          <input
            className="form-control"
            disabled={false}
            id="MySetting"
            onChange={[Function]}
            placeholder="E.g.: \\"git,smtp\\""
            type="text"
            value="git"
          />
        </LocalizedPlaceholderInput>
        <div
          className="help-text"
          data-testid="MySettinghelp-text"
        >
          <FormattedMessage
            defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
            id="admin.customization.customUrlSchemesDesc"
          >
            <span>
              Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: "http", "https", "ftp", "tel", and "mailto".
            </span>
          </FormattedMessage>
        </div>
      </div>
    </div>
  </Memo(Settings)>
</CustomURLSchemesSetting>
`;

exports[`components/AdminConsole/CustomUrlSchemeSetting renders properly when disabled 1`] = `
<CustomURLSchemesSetting
  disabled={true}
  id="MySetting"
  onChange={[MockFunction]}
  setByEnv={false}
  value={
    Array [
      "git",
      "smtp",
    ]
  }
>
  <Memo(Settings)
    helpText={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
        id="admin.customization.customUrlSchemesDesc"
      />
    }
    inputId="MySetting"
    label={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Custom URL Schemes:"
        id="admin.customization.customUrlSchemes"
      />
    }
    setByEnv={false}
  >
    <div
      className="form-group"
      data-testid="MySetting"
    >
      <label
        className="control-label col-sm-4"
        htmlFor="MySetting"
      >
        <FormattedMessage
          defaultMessage="Custom URL Schemes:"
          id="admin.customization.customUrlSchemes"
        >
          <span>
            Custom URL Schemes:
          </span>
        </FormattedMessage>
      </label>
      <div
        className="col-sm-8"
      >
        <LocalizedPlaceholderInput
          className="form-control"
          disabled={true}
          id="MySetting"
          onChange={[Function]}
          placeholder={
            Object {
              "defaultMessage": "E.g.: \\"git,smtp\\"",
              "id": "admin.customization.customUrlSchemesPlaceholder",
            }
          }
          type="text"
          value="git,smtp"
        >
          <input
            className="form-control"
            disabled={true}
            id="MySetting"
            onChange={[Function]}
            placeholder="E.g.: \\"git,smtp\\""
            type="text"
            value="git,smtp"
          />
        </LocalizedPlaceholderInput>
        <div
          className="help-text"
          data-testid="MySettinghelp-text"
        >
          <FormattedMessage
            defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
            id="admin.customization.customUrlSchemesDesc"
          >
            <span>
              Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: "http", "https", "ftp", "tel", and "mailto".
            </span>
          </FormattedMessage>
        </div>
      </div>
    </div>
  </Memo(Settings)>
</CustomURLSchemesSetting>
`;

exports[`components/AdminConsole/CustomUrlSchemeSetting renders properly when set by environment variable 1`] = `
<CustomURLSchemesSetting
  disabled={false}
  id="MySetting"
  onChange={[MockFunction]}
  setByEnv={true}
  value={
    Array [
      "git",
      "smtp",
    ]
  }
>
  <Memo(Settings)
    helpText={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
        id="admin.customization.customUrlSchemesDesc"
      />
    }
    inputId="MySetting"
    label={
      <Memo(MemoizedFormattedMessage)
        defaultMessage="Custom URL Schemes:"
        id="admin.customization.customUrlSchemes"
      />
    }
    setByEnv={true}
  >
    <div
      className="form-group"
      data-testid="MySetting"
    >
      <label
        className="control-label col-sm-4"
        htmlFor="MySetting"
      >
        <FormattedMessage
          defaultMessage="Custom URL Schemes:"
          id="admin.customization.customUrlSchemes"
        >
          <span>
            Custom URL Schemes:
          </span>
        </FormattedMessage>
      </label>
      <div
        className="col-sm-8"
      >
        <LocalizedPlaceholderInput
          className="form-control"
          disabled={true}
          id="MySetting"
          onChange={[Function]}
          placeholder={
            Object {
              "defaultMessage": "E.g.: \\"git,smtp\\"",
              "id": "admin.customization.customUrlSchemesPlaceholder",
            }
          }
          type="text"
          value="git,smtp"
        >
          <input
            className="form-control"
            disabled={true}
            id="MySetting"
            onChange={[Function]}
            placeholder="E.g.: \\"git,smtp\\""
            type="text"
            value="git,smtp"
          />
        </LocalizedPlaceholderInput>
        <div
          className="help-text"
          data-testid="MySettinghelp-text"
        >
          <FormattedMessage
            defaultMessage="Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: \\"http\\", \\"https\\", \\"ftp\\", \\"tel\\", and \\"mailto\\"."
            id="admin.customization.customUrlSchemesDesc"
          >
            <span>
              Allows message text to link if it begins with any of the comma-separated URL schemes listed. By default, the following schemes will create links: "http", "https", "ftp", "tel", and "mailto".
            </span>
          </FormattedMessage>
        </div>
        <SetByEnv>
          <div
            className="alert alert-warning"
          >
            <FormattedMessage
              defaultMessage="This setting has been set through an environment variable. It cannot be changed through the System Console."
              id="admin.set_by_env"
            >
              <span>
                This setting has been set through an environment variable. It cannot be changed through the System Console.
              </span>
            </FormattedMessage>
          </div>
        </SetByEnv>
      </div>
    </div>
  </Memo(Settings)>
</CustomURLSchemesSetting>
`;
